Package Delivery Feature Documentation (#2072)
PX4/PX4-user_guide: en/peripherals/gripper.md

# Gripper Peripherals

![High-load gripper example](../../assets/hardware/grippers/highload_gripper_example.jpg)

Grippers are a commonly used mechanism for UAV delivery applications. There are multiple types of interfaces (e.g. PWM, CAN) for different products in the market.

> ## Supported Gripper types

Reviewer: ```suggestion
## Supported Gripper Types
```

> PX4 supports following Gripper types, check the individual page to learn how to integrate them.

Reviewer: ```suggestion
PX4 supports the following Gripper types, check the individual page to learn how to integrate them.
```

> PX4 supports the following Gripper types, check the individual page to learn how to integrate them.

Reviewer: What does it mean to "support" a gripper type? I think essentially it means that for a hardware-integrated gripper you can assign the gripper output. More generally it means that you can configure it to be triggered when the gripper commmand is encountered in a mission or via a command. Right? 

So the roboclaw is not supported because you can't set it up to close when a gripper command is sent. Right?

Author: Right, ones that we can configure to be a gripper is what we support. And since there are numerous other gripper types (e.g. EPM Gripper: https://ardupilot.org/copter/docs/common-electro-permanent-magnet-V3.html), I narrowed down the scope to a 'Servo Gripper'.

And this shouldn't depend on whether it can be used in mission (right now, gripper driver is tied to `payload_deliverer` module, but it should be a separate entity actually), and ideally every gripper should be controllable via the gripper MAVLink command.

And since the gripper command doesn't work, roboclaw isn't the supported gripper as you said!

Reviewer: @junwoo091400 Thanks. So what "interface" is roboclaw? I assume it is not PWM, or it is more complicated than PWM servo? I..e I really don't want to talk about the driver per-se, more about the protocol.

> - [Servo Gripper with PWM](gripper_servo.md)

Reviewer: What about mavlink gripper support? Is it just that you don't have one to test with? What I'd expect here is that if you get the gimbal command and it isn't addressed to the FC it should be forwarded to the component that can handle it.

Author: Good point! And that's totally possible with current setup (haven't tested yet) theoretically. I'm not aware of a real MAVLink Gripper I can get off the shelf & test with. I will ask around and let you know if I get to test it :thinking: 

Or if you are aware of it, let me know as well :+1:

Reviewer: I don't know of one. I'll think on how we might integrate this.

Reviewer: Initially I'll just add a note that it is theoretically supported. Still working on this doc.

:::note
The [Roboclaw driver](../modules/modules_driver.md#roboclaw) isn't supported as a Gripper yet. However in the future it may get supported as a Gripper.
:::

> ## Gripper actuator mapping

Reviewer: ```suggestion
## Gripper Actuator Mapping
```

> ## Gripper Actuator Mapping

Reviewer: Should this go into the servo docs? I appreciate that you might later have a UAVCAN servo that is almost the same, but I think in that case it might still be good to have a separate image. Thoughts?

Reviewer: Actually I think "yes", along with the specific configuration for the servo. There might be duplication, but the servo setup will be easier to follow.

Author: I moved most of this to the Gripper Servo (along with the image, as it is PWM specific), and left generic comments only in the gripper doc

To enable the output of the Gripper, set the Function of the output port the gripper is connected to as `Gripper`.

:::note
Make sure to set the correct PWM frequency (usually commercial servo / grippers only support 50Hz)! If you mis-configure the frequency you risk damaging the gripper.
:::

![Gripper output mapping](../../assets/config/gripper/qgc_gripper_output_setup.png)

> ## Griper action joystick button mapping in QGC

Reviewer: ```suggestion
## Gripper Action Joystick Button Mapping in QGC
```

> QGC provides the Button mapping interface, which allows to map delivery mechanism actions to specific buttons of the [Joystick](../config/joystick.md) connected to QGC.

Reviewer: Can an RC controller switch be mapped in http://docs.px4.io/main/en/config/flight_mode.html#flight-mode-selection

I'm guessing not, but we should make that clear somewhere.

### Buttons Mapping

Navigate by clicking **"QGC Logo (upper-left)" > "Vehicle Setup" > "Joystick" > "Button Assignment"** tab. There are Gripper Open and Gripper Close actions available to map to a button.

![Gripper action mapping](../../assets/config/gripper/qgc_gripper_actions_joystick.png)

You can test the actions by clicking on the buttons each and checking Gripper's movements. If the gripper doesn't move as expected, please make sure that the Gripper integration was set up properly.

If you press "Grab" button while the gripper is opening, it will automatically abort releasing behavior and go to the Close position, effectively cancelling the release command.

:::note
Note that currently in PX4 the Gripper isn't supported as a standalone driver, but is instead tied to the [payload delivery feature](../advanced_features/package_delivery.md).

> Therefore unless you [configure the gripper for package delivery](#setting-up-gripper-for-package-delivery), gripper will not be functioning!

Reviewer: To confirm, the package delivery setup is what links the mavlink gripper command and the gripper actuator output right?

So if you don't enable the payload delivery feature the system might get the mavlink command to grip/ungrip but won't pass that through to the servo?

Author: Related to the [comment above](https://github.com/PX4/PX4-user_guide/pull/2072#discussion_r1020336768), gripper's functionality shouldn't be dependent on the package delivery feature ultimately.

And so for now, package delivery setup is a necessary evil for getting gripper to work, but that will be resolved in the future.

Reviewer: Yes, I get that. I was hoping you would confirm the "detail". But I think I'm correct.

:::

## Setting up Gripper for Package Delivery

Gripper is commonly used for features like [Package Delivery](../advanced_features/package_delivery.md). This section describes how to setup the Payload Delivery for Gripper.

> ### Note on using Joystick button action during package delivery mission

Reviewer: ```suggestion
### Note on Using Joystick Button Action During Package Delivery Mission
```

If you forcefully command the gripper to 'Close' position while the package delivery is happening (gripper open action), gripper won't be able to finish the open action, and the mission will come to a halt for a payload delivery mission item timeout, then resume the mission.

> ### Enable payload delivery feature (Gripper)

Reviewer: ```suggestion
### Enable Payload Delivery Feature (Gripper)
```

Set [`PD_GRIPPER_EN`](../advanced_config/parameter_reference.md#PD_GRIPPER_EN) parameter to 1 (reboot required after change).

> ### Enabling Pre-arm mode

Reviewer: ```suggestion
### Enabling Pre-arm Mode
```

Set [`COM_PREARM_MODE`](../advanced_config/parameter_reference.md#COM_PREARM_MODE) to `Always`.

### Profiling the Gripper

Some mechanism specific settings need to be added to make sure the system is aware of the mechanism's physical properties. Since every gripper works differently (speed / range of motion / behavior).

### Gripper Mechanism Type

Set the [`PD_GRIPPER_TYPE`](../advanced_config/parameter_reference.md#PD_GRIPPER_TYPE) to the type of Gripper you have.

Note that for now we only support a [Servo Gripper](gripper_servo.md), so you will be setting the value to `Servo`.

### Gripper Actuation Time

You need to specify an actuation time, which specifies the time it takes to open or close the gripper. As most grippers don't have a sensor to detect successful actuation (close/opening), the payload delivery feature will rely on this value to estimate gripper's position.

> To measure gripper actuation time, go to the MAVLink Shell in AMC and execute the following code while the drone is on a bench and the propellers are removed:

Reviewer: ```suggestion
To measure gripper actuation time, go to the [MAVLink Shell](../debug/mavlink_shell.md) in QGC  and execute the following code while the drone is on a bench and the propellers are removed:
```

`> payload_deliverer gripper_test`

> Then observe how long it takes for the gripper to go from open position to closed position, and vise versa for opening. Note down the actuation time for both cases and choose the maximum measurement for the actuation time.

Reviewer: ```suggestion
Then observe how long it takes for the gripper to go from open position to closed position, and vice versa for opening. Note down the actuation time for both cases and choose the maximum measurement for the actuation time.
```

As an alternative test method, you can utilize the [Joystick capability](#griper-action-joystick-button-mapping-in-qgc) to trigger gripper open and close actions as well.

:::note
If you get an error message such as "[payload_deliverer] not running", make sure that you have gone through the setup procedures above and either run `payload_deliverer start` command in the Nuttx shell
:::

After measuring the opening and closing actuation time, set [PD_GRIPPER_TO](../advanced_config/parameter_reference.md#PD_GRIPPER_TO) to the longer timeout of the two.

> ### Mission timeout

Reviewer: ```suggestion
### Mission Timeout
```

# Gripper Peripherals

Grippers are mechanical devices that can be integrated with an unmanned vehicle to hold (grip) and release payloads.

PX4 allows grippers to triggered automatically in [Payload Delivery Missions](../flying/package_delivery_mission.md) or manually using a Joystick.

![High-load gripper example](../../assets/hardware/grippers/highload_gripper_example.jpg)

> :::note

Reviewer: FYI new note. I realised that if used as a generic actuator a gripper is not a gripper. So this links to the actuator docs - if someoen wants to do that they know enough now to find it. The rest of this doc is specific to your solution.

Author: Right, I could add a section on using Gripper as a 'generic actuator', as that is probably the most common use case!

Author: Added in https://github.com/PX4/PX4-user_guide/pull/2072/commits/8cb6995af87a95eeadbb32f22a5dad2be1c2b1a2

Reviewer: A gripper is a gripper (as you have documented). A generic actuator that happens to be mapped to a gripper is not a gripper :-). It's enough to have this note. I have tidied it up.

A gripper can also be configured as a [generic RC or MAVLink actuator](../payloads/README.md#actuator-control-with-rc).
This allows it to be used manually via an RC Controller, but not in missions or on a Joystick.
:::

## Supported Grippers

There are many different gripper mechanisms ("jaws", "fingers", "electromagnets") and interfaces (PWM, CAN, MAVLink, and so on).

PX4 supports grippers that have simple triggers to hold and release, and that use the following interfaces (see linked documents for details):

- [PWM Servo Gripper](gripper_servo.md) - Grippers connected to autopilot PWM outputs

> - **MAVLink Gripper** (Untested) - Grippers that support the [MAV_CMD_DO_GRIPPER](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_GRIPPER) MAVLink command.

Reviewer: Note, so we need to talk about setup and limitations, but I think this is good for now. Perhaps a topic as a follow up when we have one we can test :-)

Author: Yep especially this hasn't been tested :thinking:

> :::note

Reviewer: How does Roboclaw interface work?

Author: For now it just takes in `actuator_controls_0` topic, and directly maps the throttle and yaw output of the flight controller into the drive & turn commands of the Roboclaw. So I am not sure if it would work so well now, and it probably needs to be refactored to use Control allocation Actuators architecture, right @bkueng ?

Reviewer: What I mean is "does it have a PWM interface"? How is it connected? I'd like to say something like "Roboclaw is not yet supported because XXX bus is not integrated with YYYY"

Reviewer: @junwoo091400 I have updated the whole of the gripper and gripper servo docs. For those docs only question is this one ^^^

The [Roboclaw driver](../modules/modules_driver.md#roboclaw) isn't supported as a Gripper yet.
However in the future it may get supported as a Gripper.
:::

> ## Using a Gripper

Reviewer: Note, new section - contains all the "using" information.

If you do this in a mission while the release is actually happening then the [delivery will be cancelled](../flying/package_delivery_mission.md#manual-control-of-gripper-in-missions).

MAVLink applications, such as ground stations, can also control the gripper using the [MAV_CMD_DO_GRIPPER](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_GRIPPER) MAVLink command.

## PX4 Configuration

### Package Delivery Configuration

PX4 Gripper support is tied to the package delivery feature, which must be enabled and configured in order to be able to use a gripper.

1. Set [`PD_GRIPPER_EN`](../advanced_config/parameter_reference.md#PD_GRIPPER_EN) parameter to 1 (reboot required after change).
1. Set [`PD_GRIPPER_TYPE`](../advanced_config/parameter_reference.md#PD_GRIPPER_TYPE) to match your Gripper.
   For example, set to `Servo` for a [Servo Gripper](gripper_servo.md).

### Gripper Actuator Mapping

To enable the output of the Gripper, set the Function of the output port where the gripper is connected to as `Gripper` in the [Actuators](../config/actuators.md#actuator-outputs) tab.

> <!-- Replace this with actuator diagram -->

Reviewer: @junwoo091400 Have we got a screenshot of a gripper setup in control allocation? Would be good here.

Author: Added!

### QGC Joystick Configuration

QGroundControl [Joystick](../config/joystick.md) configuration allows you to map gripper actions to Joystick buttons, after which you will be open and close the gripper manually.

Open the Joystick configuration by selecting: **QGC Logo (upper-left) > Vehicle Setup > Joystick > Button Assignment** tab.
Select `Gripper Open` and `Gripper Close` actions for your desired joystick buttons

![Gripper action mapping](../../assets/config/gripper/qgc_gripper_actions_joystick.png)

You can test the actions by clicking on the mapped buttons and checking for gripper movement.
If the gripper doesn't move as expected check the package delivery configuration and actuator mapping are set up properly.

### Enable Pre-ARM Mode

We need to set a parameter to enable [pre-arming](../advanced_config/prearm_arm_disarm.md).
This keeps the motors disabled but allows the gripper to be opened and closed for attaching the payload (avoiding potential danger from spinning propellers).

> Set [`COM_PREARM_MODE`](../advanced_config/parameter_reference.md#COM_PREARM_MODE) to `Always`.

Reviewer: Do we need a caveat or warning, and if so what? I.e. if it is completely safe to always set prearm mode, then why isn't this always enabled or would default-on. My thinking is that there are case where prearm might be dangerous.

Author: I have asked @MaEtUgR about it!

Reviewer: Thanks @junwoo091400 - What did he say?

Reviewer 2: I said I so far always enabled prearm mode especially for preflight surface checks. I'm guessing it was previously disabled by default such that if you accidentally plug a motor on a servo pin the chances are lower that it spins up. But that assumption is flawed since servos usually have 1500 as disarmed values. Also assignment of actuators is now much better with control allocation. So the only remaining thing I see is to save the servos from extra work while the vehicle is powered on the bench 🤷‍♂️ But a fixed-wing pro like @sfuhrer might have more insights.

Reviewer: @MaEtUgR Thanks. IMO we don't particularly need a warning. As you say, prearm is only unsafe if somehow motors or "dangerous actuators" can move in this state. So possibly risky on first setup, but not for use once properly configured.

Reviewer 2: Yes, exactly 👍

Reviewer 3: Context: https://github.com/PX4/PX4-Autopilot/pull/17309#issuecomment-992205444
TL;DR: I think we shouldn't just change the default to "Always", but stay on the safe side.
A user then should make the deliberate decision (after setting up the vehicle the first time and testing the wiring) whether one wants to have the servos already enabled by without arming or in prearm mode. But in general "arming" should be active for all actuators, not just the motors.

Reviewer: Fair enough. I think at this point you've probably already verified your motors and servos are set up properly and prearm is safe. So no need for a warning. No harm on being on the safe side for the setting either.

> ### Configure Gripper Actuation Time

Reviewer: I merged the profiling topic into this, but I have not re-written it yet. So consider subedit "up to here".
